### PR TITLE
Update of the HTO skript

### PR DESCRIPTION
**Pull request by andpet0101**
This small pull request updates the HTO Skript and makes it more stable for production. The changes are mostly cosmetic. Most importantly normalisation for UMAP is now done with Seurat's standard LogNormalize approach.  This is followed by UMAP for visualisation. Standard clustering is done as well and includes doublets and negatives. Distribution of cells in clusters for each individual HTO+negatives/doublets (and vice versa) as well as QC (counts, features) parameters for each cluster and HTO+negatives/doublets are very informative for cell quality and could be included in filtering later on.

IMPORTANT: Please do the PR only once the other PR for the DEGs is done. There are some minor changes which also apply to the main script.

### CODE REVIEW

File: R/functions_io.R [hunk 1]
```diff
@@ -149,7 +149,7 @@ ReadCountsTable = function(counts_table, project="SeuratProject", row_name_colum
     # Include cell and feature metadata
     c = samples_to_process[[s]] # cells to include for sample
     f = feature_types[1] # feature type
-    a = feature_type_to_assay_name[f] # assay name
+    a = unname(feature_type_to_assay_name[f]) # assay name
```
Comment (ktrns) [inline R/functions_io.R:162]: What is `unname` doing here?

File: R/functions_io.R [hunk 2]
```diff
@@ -195,10 +195,11 @@ ReadCountsTable = function(counts_table, project="SeuratProject", row_name_colum
 #' @param convert_row_names Named vector for converting the row names obtained from features.tsv.gz (e.g. from Ensembl ids to gene symbols). Does not need to contain all row names. Can be NULL in which case row names are not converted.
 #' @param feature_type_column Name or index of the column which should be used for feature types (3).
 #' @param feature_type_to_assay_name How should the assays for the different feature type be named? Default is: "Gene expression" = "RNA","Antibody Capture" = "ADT","CRISPR Guide Capture" = "Crispr", "ERCC" = "ERCC" and "Custom" = "Custom". Also sets the order in which the assays are loaded.
-#' @param hto_names If Antibody Capture data, is used hashtags for multiplexing, a vector with feature names to be used as hashtags. If a named vector is provided, the hashtags will be renamed accordingly.
+#' @param hto_names If Antibody Capture data is used as hashtags for multiplexing, a vector with feature names to be used as hashtags. If a named vector is provided, the hashtags will be renamed accordingly.
```
Comment (ktrns) [inline R/functions_io.R]: are used

File: R/functions_io.R [hunk 3]
```diff
@@ -195,10 +195,11 @@ ReadCountsTable = function(counts_table, project="SeuratProject", row_name_colum
...
+#' @param hto_regex If Antibody Capture data is used as hashtags for multiplexing, a regular expression to identify the hashtag (e.g. HashTag).
 #' @param return_samples_as_datasets If there are multiple samples in the dataset: return each as separate dataset ("TRUE").
 #' @return A Seurat object.
-ReadSparseMatrix = function(path, project="SeuratProject", row_name_column=2, convert_row_names=NULL, feature_type_column=3, feature_type_to_assay_name=NULL, hto_names=NULL) {
+ReadSparseMatrix = function(path, project="SeuratProject", row_name_column=2, convert_row_names=NULL, feature_type_column=3, feature_type_to_assay_name=NULL, hto_names=NULL, hto_regex=NULL) {
   library(magrittr)
   
   # defaults
```
Comment (ktrns) [inline R/functions_io.R]: Can we stick to comments that start with capital letters?

Comment (ktrns) [reply R/functions_io.R]: It seems that in the HTO script we don't really stick to it at all. I can also go through it once the PRs are done.

File: R/functions_io.R [hunk 4]
```diff
@@ -240,6 +241,12 @@ ReadSparseMatrix = function(path, project="SeuratProject", row_name_column=2, co
   }
 
   # special case: if hashtags are specified and if there is a Antibody Capture data, then split and add as separate assay
+  # if a regular expression is defined for HTO, overwrite hto_names 
+  if ("Antibody Capture" %in% names(feature_data) && !is.null(hto_regex) && nchar(hto_regex)>0) {
+    hto_names = grep(pattern=hto_regex, x=rownames(feature_data[["Antibody Capture"]]), v=TRUE)
+    if (length(hto_names)==0) futile.logger::flog.error("Could not find HTO names with the 'hto_regex' argument '%s'!", hto_regex)
```
Comment (ktrns) [inline R/functions_io.R]: So ... I think we said some time ago, we drop the `futile.logger`?

File: R/functions_io.R [hunk 5]
```diff
@@ -338,8 +345,9 @@ ReadSparseMatrix = function(path, project="SeuratProject", row_name_column=2, co
 #' @param feature_type_to_assay_name How should the feature types for the different assays be named? Default is: "RNA" = "Gene expression","ADT" = "Antibody Capture" ,"HTO" = "Antibody Capture","Crispr" = "CRISPR Guide Capture", "ERCC" = "ERCC" and "Custom" = "Custom". Also sets the order in which the feature types are written.
 #' @param include_cell_metadata_cols Vector with names of cell metadata columns to be written to the metadata.tsv.gz. Can be NULL in which case no file is created.
 #' @param include_feature_metadata_cols Vector with names of feature metadata columns to be written. Can be NULL in which case only the first three columns (feature id, feature symbol and feature type) are written.
+#' @param metadata_prefix Prefix for cell metadata column names. If NULL, no prefix. Separator must be included.
```
Comment (ktrns) [inline R/functions_io.R]: I don't get this one. What do you mean by "Separator must be included."?

If NULL, no prefix is added.

File: scrnaseq.Rmd [hunk 6]
```diff
@@ -183,6 +187,14 @@ file_exists = purrr::map_lgl(git_files_to_source, file.exists)
 if (any(!file_exists)) stop(paste("The following files could not be found:",paste(git_files_to_source[!file_exists], collapse=", "), ". Please check the git directory at '", param$path_to_git, "'.!"))
 invisible(purrr::map(git_files_to_source, source))
 
+# Debugging mode: 
+# 'default_debugging' for default, 'terminal_debugger' for debugging without X11, 'print_traceback' for non-interactive sessions 
+switch (param$debugging_mode, 
+        default_debugging=on_error_default_debugging(),
```
Comment (ktrns) [inline scrnaseq.Rmd:236]: What is the `switch` doing?

File: test_datasets/10x_pbmc_hto_GSE108313/download.R [hunk 7]
```diff
@@ -1,13 +1,25 @@
 # Downloaded 10X data set, 'pbmc_hto_mtx.rds' and 'pbmc_umi_mtx.rds'
 # https://www.dropbox.com/sh/ntc33ium7cg1za1/AAD_8XIDmu4F7lJ-5sp-rGFYa?dl=0
+#
```
Comment (ktrns) [inline test_datasets/10x_pbmc_hto_GSE108313/download.R:3]: Just wondering, was there anything wrong here or you just wanted to improve the script generally?

File: test_datasets/10x_pbmc_hto_GSE108313/download.R [hunk 8]
```diff
@@ -31,31 +43,20 @@ pbmc_umi = pbmc_umi[,joint_barcodes]
...
-# Add Ensembl IDs to features.tsv
-genes$Ensembl = c(genes_ensembl[genes$GeneSymbol], hto_names)
+feat = data.frame(V1=genes_ensembl[rownames(pbmc_umi)], V2=rownames(pbmc_umi), V3="Gene Expression")
+feat = rbind(feat, data.frame(V1=rownames(pbmc_hto), V2=rownames(pbmc_hto), V3="Antibody Capture"))
 
-# Write genes to features.tsv
-write.table(genes, file="counts/features.tsv", sep="\t", row.names=FALSE, col.names=FALSE, quote=FALSE)
+# write matrix.mtx.gz, barcodes.tsv.gz and features.tsv.gz
+mh = file.path("counts","matrix.mtx")
+Matrix::writeMM(pbmc_counts,file=mh)
```
Comment (ktrns) [inline test_datasets/10x_pbmc_hto_GSE108313/download.R]: Take care of the spaces in this script.

File: test_datasets/10x_pbmc_hto_GSE108313/download.R [hunk 9]
```diff
@@ -31,31 +43,20 @@ pbmc_umi = pbmc_umi[,joint_barcodes]
...
-system("gzip counts/barcodes.tsv")
-system("gzip counts/features.tsv")
-system("gzip counts/matrix.mtx")
+bh = gzfile(file.path("counts", "barcodes.tsv.gz"), open="wb")
+write(colnames(pbmc_counts), file=bh)
+close(bh)
 
-# Remove genes.tsv
-# This is necessary, so Seurat recognises the CellRanger version
-system("rm counts/genes.tsv")
```
Comment (ktrns) [inline test_datasets/10x_pbmc_hto_GSE108313/download.R:57]: I vaguely remember that removing `genes.tsv` was necessary. In your case, is it not written in the first place?

File: scrnaseq_hto.Rmd [hunk 10]
```diff
@@ -16,17 +16,30 @@ params:
...
-   #     param$hto.names = setNames(c("NameA", "NameB", "NameC"), c("HTO1", "HTO2", "HTO3"))
+   #    - without names: 'hto.names: c("HTO1", "HTO2", "HTO3")'
+   #    - with names: 'hto.names: !r setNames(c("NameA", "NameB", "NameC"), c("HTO1", "HTO2", "HTO3"))'
+   # 
    hto_names: !r setNames(c("htoA","htoB","htoC","htoD","htoE","htoF","htoG","htoH"), c("htoA","htoB","htoC","htoD","htoE","htoF","htoG","htoH"))
    
+   # HTO regular expression (alternative to HTO names)
+   #   Identify HTOs via a regular expression. Names cannot be provided here.
+   #   If specified will overrule hto_names
+   #hto_regex: "^hto"
```
Comment (ktrns) [inline scrnaseq_hto.Rmd:39]: What will `"^hto"` do?

File: scrnaseq_hto.Rmd [hunk 11]
```diff
@@ -16,17 +16,30 @@ params:
...
+   
+   # HTO normalisation: CLR (default) or LogNorm
+   norm: "CLR"
```
Comment (ktrns) [inline scrnaseq_hto.Rmd]: Why did you switch the default?

File: scrnaseq_hto.Rmd [hunk 12]
```diff
@@ -55,31 +81,94 @@ library(magrittr) # %>% operator
...
 ```
 
-```{r hto_colours}
-hto_colours = list()
-hto_colours$col_hto_global = ggsci::pal_npg()(3)
-hto_colours$col_hto_all = ggsci::pal_npg()(factorial(length(params$hto_names)) + 1)
-hto_colours$col_hto_collapsed = ggsci::pal_npg()(length(params$hto_names) + 2)
-hto_levels = c("Negative", "Doublet", params$hto_names[length(params$hto_names):1])
-names(hto_colours$col_hto_collapsed) = hto_levels
+```{r preparation_and_initial_checks}
```
Comment (ktrns) [inline scrnaseq_hto.Rmd:102]: What is your motivation for all these changes and all the additional code? To make it more in line with the main script?

File: scrnaseq_hto.Rmd [hunk 13]
```diff
@@ -133,39 +229,49 @@ This section of the report shows how cells are assigned to their sample-of-origi
 
 ## Normalisation of HTO counts
 We start the analysis by normalising raw HTO counts. HTO counts for each cell are divided by the total counts for that cell and multiplied by 10,000. This is then natural-log transformed. 
-```{r hto_norm}
-sc = Seurat::NormalizeData(sc, assay="HTO", normalization.method="LogNormalize")
+```{r hto_normalisation}
+
+# HTO
```
Comment (ktrns) [inline scrnaseq_hto.Rmd:240]: How about `param$norm` is "CLR" or "LogNormalize" and you skip the if else, and just call it with `param$norm`?

File: scrnaseq_hto.Rmd [hunk 14]
```diff
@@ -133,39 +229,49 @@ This section of the report shows how cells are assigned to their sample-of-origi
...
+  sc = Seurat::NormalizeData(sc, assay="HTO", normalization.method="CLR", verbose=FALSE)
+} else if (params$norm=="LogNorm") {
+  # alternative: LogNorm
+  sc = Seurat::NormalizeData(sc, assay="HTO", normalization.method="LogNormalize", verbose=FALSE)
+}
 
-# alternative: CLR
-# sc = NormalizeData(sc, assay="HTO", normalization.method="CLR") 
+# RNA
+sc = Seurat::NormalizeData(sc, assay="RNA", normalization.method = "LogNormalize", scale.factor=10000, verbose=FALSE)
```
Comment (ktrns) [inline scrnaseq_hto.Rmd]: Why do you do this here?

File: scrnaseq_hto.Rmd [hunk 15]
```diff
@@ -179,7 +285,86 @@ p = p1 + p2 +
 p
 ```
 
-# Visualisation of raw and normalised HTO data {.tabset}
+```{r hto_demux_debug, eval=FALSE, include=FALSE}
```
Comment (ktrns) [inline scrnaseq_hto.Rmd]: Mmmh .... does it make sense to have this kind of debugging code in the script? If HTODemux changes, this is still old ...

File: scrnaseq_hto.Rmd [hunk 16]
```diff
@@ -189,15 +374,15 @@ hto_t_raw_pseudo = hto_t_raw + 1
 hto_t_norm = Seurat::GetAssayData(sc, assay="HTO", slot="data") %>% 
   as.data.frame %>% t %>% as.data.frame
 
-p1 = ggplot(hto_t_raw_pseudo %>% tidyr::pivot_longer(tidyr::everything()), aes(x=name, y=value, fill=name)) + 
+p1 = ggplot(hto_t_raw_pseudo %>% tidyr::gather(key="name"), aes(x=name, y=value, fill=name)) +
```
Comment (ktrns) [inline scrnaseq_hto.Rmd]: Out of interest: what is `gather` doing here compared to `pivot_longer`?

File: scrnaseq_hto.Rmd [hunk 17]
```diff
@@ -234,6 +419,7 @@ p
 
 ```{r hto_plot_norm_scatter, fig.height=10}  
 n = DfAllColumnCombinations(x=hto_t_norm, cell_classification=sc$hash.ID)
+n = n[n$value1>0 & n$value2>0,]
```
Comment (ktrns) [inline scrnaseq_hto.Rmd]: Space

File: scrnaseq_hto.Rmd [hunk 18]
```diff
@@ -258,23 +444,40 @@ p = patchwork::wrap_plots(p_list, ncol = 2) +
...
-Lastly, we compare the number of features between classified cells. 
+Lastly, we compare the counts and number of features between classified cells. 
 ```{r hto_plot_features}
 # Number of features in the different cells
 nfeature_metrics = grep("_HTO", 
                         grep("nFeature_", colnames(sc[[]]), v=TRUE),
                         v=TRUE,
                         invert=TRUE)
-p = VlnPlot(sc, features=nfeature_metrics, idents=levels(Seurat::Idents(sc)), pt.size=0) + 
+ncounts_metrics = grep("_HTO",
```
Comment (ktrns) [inline scrnaseq_hto.Rmd:375]: Not quite sure what this is, but will see once I run the script.

File: scrnaseq_hto.Rmd [hunk 19]
```diff
@@ -289,8 +492,38 @@ sc
 # Preliminary visualisation of demultiplexed RNA data
 This section of the report provides first insights into your RNA dataset based on a preliminary pre-processing of the RNA data using the standard scRNA-seq workflow. 
 ```{r prelim_analysis, results='hide'}
-sc.all = prelim.analysis(sc=sc.all, mt=params$mt_names, pc_n=10)
-sc = prelim.analysis(sc=sc, mt=params$mt_names, pc_n=10)
+# normalise and scale sc
```
Comment (ktrns) [inline scrnaseq_hto.Rmd]: I have seen the same code further up already? (norm, var, scale)

File: scrnaseq_hto.Rmd [hunk 20]
```diff
@@ -289,8 +492,38 @@ sc
 # Preliminary visualisation of demultiplexed RNA data
 This section of the report provides first insights into your RNA dataset based on a preliminary pre-processing of the RNA data using the standard scRNA-seq workflow. 
 ```{r prelim_analysis, results='hide'}
-sc.all = prelim.analysis(sc=sc.all, mt=params$mt_names, pc_n=10)
-sc = prelim.analysis(sc=sc, mt=params$mt_names, pc_n=10)
```
Comment (ktrns) [inline scrnaseq_hto.Rmd:293]: So you skipped the function, and rather write the code in the main script? Is that the intention?

File: scrnaseq_hto.Rmd [hunk 21]
```diff
@@ -324,34 +559,137 @@ p = Seurat::DimPlot(sc, reduction="umap", group.by="hash.ID", cols=hto_colours$c
...
+### Before filtering (each hto)
+```{r umap_cluster_before, fig.height=sc_all_fig_height}
+p = Seurat::DimPlot(sc.all, reduction="umap", group.by="seurat_clusters", split.by="hash.ID", ncol=2, cols=sc_all_col_clusters) + 
+  AddStyle(title="UMAP for each HTO, including doublets and negatives", 
+           legend_title="Cluster", 
+           legend_position="right")
+p
+```
+
+* Important: cluster colors are not neccessarily the same before and after filtering!
```
Comment (ktrns) [inline scrnaseq_hto.Rmd]: Cluster (capital C)

File: scrnaseq_hto.Rmd [hunk 22]
```diff
@@ -324,34 +559,137 @@ p = Seurat::DimPlot(sc, reduction="umap", group.by="hash.ID", cols=hto_colours$c
...
+### After filtering (each hto)
+```{r umap_cluster_after, fig.height=sc_fig_height}
+p = Seurat::DimPlot(sc, reduction="umap", group.by="seurat_clusters", split.by="hash.ID", ncol=2, cols=sc_col_clusters) + 
+  AddStyle(title="UMAP for each HTO, singlets only", 
+           legend_title="Cluster", 
+           legend_position="right")
+p
+```
+
+* Important: cluster colors are not neccessarily the same before and after filtering!
```
Comment (ktrns) [inline scrnaseq_hto.Rmd]: Cluster

File: scrnaseq_hto.Rmd [hunk 23]
```diff
@@ -324,34 +559,137 @@ p = Seurat::DimPlot(sc, reduction="umap", group.by="hash.ID", cols=hto_colours$c
...
+           legend_position="right")  
+
+p = patchwork::wrap_plots(list(p1, p2), ncol=2) & theme(legend.position="bottom") 
+p = p + patchwork::plot_annotation("Distribution of HTO and clusters") + 
+  patchwork::plot_layout(guides = "collect")
+p
+
+```
+
+Here it also helps to inspect RNA counts and features in these clusters by hto.
```
Comment (ktrns) [inline scrnaseq_hto.Rmd]: What do you mean by "in these clusters by hto"?

File: scrnaseq_hto.Rmd [hunk 24]
```diff
@@ -324,34 +559,137 @@ p = Seurat::DimPlot(sc, reduction="umap", group.by="hash.ID", cols=hto_colours$c
...
+p = Seurat::DimPlot(sc, reduction="umap", group.by="seurat_clusters", split.by="hash.ID", ncol=2, cols=sc_col_clusters) + 
+  AddStyle(title="UMAP for each HTO, singlets only", 
+           legend_title="Cluster", 
+           legend_position="right")
+p
+```
+
+* Important: cluster colors are not neccessarily the same before and after filtering!
+
+## Distribution of cells in HTO and in clusters
```
Comment (ktrns) [inline scrnaseq_hto.Rmd:506]: That's nice

**Review by ktrns — CHANGES_REQUESTED** [review]
Hi & thanks for the update! 

I have added a few questions that we can talk about later. 
Thanks, 
Katrin

File: scrnaseq_hto.Rmd (hunk 20 above)
Comment (ktrns) [reply scrnaseq_hto.Rmd:293]: Remove the function if it isn't needed any more.

File: scrnaseq_hto.Rmd (hunk 23 above)
Comment (ktrns) [reply scrnaseq_hto.Rmd]: Here it also helps to inspect RNA counts and features per cluster and hto.

File: R/functions_io.R (hunk 2 above)
Comment (andpet0101) [reply R/functions_io.R]: Done

File: R/functions_io.R (hunk 4 above)
Comment (andpet0101) [reply R/functions_io.R]: I replaced it with `stop(sprintf("Could not find HTO names with the 'hto_regex' argument '%s'!", hto_regex))`. Use of `sprintf` makes it more readable then `paste`.

File: R/functions_io.R (hunk 5 above)
Comment (andpet0101) [reply R/functions_io.R]: I think we take care of this when we do the output (so after this PR)

File: scrnaseq.Rmd (hunk 6 above)
Comment (andpet0101) [reply scrnaseq.Rmd:236]: Explained

File: test_datasets/10x_pbmc_hto_GSE108313/download.R (hunk 7 above)
Comment (andpet0101) [reply test_datasets/10x_pbmc_hto_GSE108313/download.R:3]: The script is now fully automatic

File: test_datasets/10x_pbmc_hto_GSE108313/download.R (hunk 8 above)
Comment (andpet0101) [reply test_datasets/10x_pbmc_hto_GSE108313/download.R]: Done

File: test_datasets/10x_pbmc_hto_GSE108313/download.R (hunk 9 above)
Comment (andpet0101) [reply test_datasets/10x_pbmc_hto_GSE108313/download.R:57]: No it worked

File: scrnaseq_hto.Rmd (hunk 10 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd:39]: All features starting with `hto` will be consered HTO.

File: scrnaseq_hto.Rmd (hunk 11 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd]: Back to LogNorm

File: scrnaseq_hto.Rmd (hunk 13 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd:240]: Done

File: scrnaseq_hto.Rmd (hunk 12 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd:102]: Yes

File: scrnaseq_hto.Rmd (hunk 14 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd]: Moved it further below

File: scrnaseq_hto.Rmd (hunk 15 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd]: Removed it

File: scrnaseq_hto.Rmd (hunk 16 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd]: Went back to `pivot_longer`

File: scrnaseq_hto.Rmd (hunk 17 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd]: Done

File: scrnaseq_hto.Rmd (hunk 18 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd:375]: Shows counts and features for the HTOs

File: scrnaseq_hto.Rmd (hunk 19 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd]: It is now here

File: scrnaseq_hto.Rmd (hunk 20 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd:293]: Removed function and file

File: scrnaseq_hto.Rmd (hunk 22 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd]: Done

File: scrnaseq_hto.Rmd (hunk 21 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd]: Hmpf done

File: scrnaseq_hto.Rmd (hunk 23 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd]: Done

File: scrnaseq_hto.Rmd (hunk 24 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd:506]: Thank you

File: R/functions_io.R (hunk 4 above)
Comment (ktrns) [reply R/functions_io.R]: Why not use a function like we use for messages and warnings?

Comment (andpet0101) [reply R/functions_io.R]: Can we take care of this after the release - i mean it is not really related to HTO but something more general?

File: R/functions_io.R (hunk 5 above)
Comment (ktrns) [reply R/functions_io.R]: Can you for now rewrite the comment, I don't find it very clear. 
Prefix for cell metadata column names. If NULL, no prefix is prepended. The prefix should end with a separator character that separates the prefix from the original column name.

File: R/functions_io.R [hunk 25]
```diff
@@ -1,5 +1,12 @@
+
+
+#' Reads one or more counts tables (e.g. provided by SmartSeq-2) and converts them into Seurat objects.
```
Comment (ktrns) [inline R/functions_io.R]: What is this about?

File: R/functions_io.R (hunk 5 above)
Comment (andpet0101) [reply R/functions_io.R]: okay I will do this

File: scrnaseq.Rmd [hunk 26]
```diff
@@ -31,6 +31,8 @@ options(stringsAsFactors=FALSE,
 reticulate_python3_path = unname(Sys.which("python3"))
 Sys.setenv(RETICULATE_PYTHON = reticulate_python3_path)
 
+Sys.setenv(RETICULATE_PYTHON = "C:/Program Files/Python38/python.exe")
```
Comment (ktrns) [inline scrnaseq.Rmd]: This looks like we have 2 rows setting RETICULATE_PYTHON?

File: scrnaseq.Rmd [hunk 27]
```diff
@@ -169,20 +171,31 @@ param$downsample_cells_n = NULL
 
 # Path to git repository
 param$path_to_git = "."
+
+# Debugging mode: 
+# 'default_debugging' for default, 'terminal_debugger' for debugging without X11, 'print_traceback' for non-interactive sessions 
+param$debugging_mode = "default_debugging"
```
Comment (ktrns) [inline scrnaseq.Rmd:216]: We need to add it to the README.

File: scrnaseq_hto.Rmd [hunk 28]
```diff
@@ -7,26 +7,40 @@ geometry: margin=2cm
 output:
    html_document:
     toc: true
+    toc_depth: 4
```
Comment (ktrns) [inline scrnaseq_hto.Rmd:10]: For later: If we stick to the floating toc, we can introduce it here as well.

File: scrnaseq.Rmd (hunk 26 above)
Comment (andpet0101) [reply scrnaseq.Rmd]: Its custom and I will delete it

File: R/functions_io.R (hunk 4 above)
Comment (ktrns) [reply R/functions_io.R]: It came up due to a change you introduced in the hto branch... I will note it down, we'll take care of it later.

File: scrnaseq_hto.Rmd (hunk 28 above)
Comment (andpet0101) [reply scrnaseq_hto.Rmd:10]: Already done

File: R/functions_io.R (hunk 25 above)
Comment (andpet0101) [reply R/functions_io.R]: I deleted it

File: scrnaseq.Rmd (hunk 27 above)
Comment (andpet0101) [reply scrnaseq.Rmd:216]: Can you add it? Or how do I include your changes to the README?

Comment (ktrns) [reply scrnaseq.Rmd:216]: I'll add it.

**Pull request merged**